pkg/util/docker/dockerfile/builder/imageprogress/progress.go

```diff
@@ -132,10 +152,32 @@ type imageProgressWriter struct {
 	layersChangedFn        func(report, report) bool
 }
 
-func (w *imageProgressWriter) readProgress() error {
+func (w *imageProgressWriter) ReadFrom(reader io.Reader) (int64, error) {
+	decoder := json.NewDecoder(reader)
+	return 0, w.readProgress(decoder)
+}
+
+func (w *imageProgressWriter) Write(data []byte) (int, error) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	if w.internalWriter == nil {
+		var pipeIn *io.PipeReader
+		pipeIn, w.internalWriter = io.Pipe()
+		decoder := json.NewDecoder(pipeIn)
+		go func() {
+			err := w.readProgress(decoder)
+			if err != nil {
+				pipeIn.CloseWithError(err)
+			}
+		}()
+	}
+	return w.internalWriter.Write(data)
+}
```

Comment:
are ReadFrom and Write only implemented so you can test this with ioutil.Copy?  Is this something we'd expect to have a real use for?